Dawn of the vfio 2.0 episode 2 reborn global offensive infinte u vite a new hope remastered enhanced edition rebooted electric boogaloo on new tides

Ediacarium opened the pull request:
Fixes #105 

Closes #84 (obsolete then)

Ediacarium committed:
Created a small suid helper binary to help with #105

Ediacarium committed:
Added gitignore

Ediacarium committed:
Install vfio-ubind to /usr/lib/windows-gaming

Ediacarium committed:
Adjusted permissons

Ediacarium committed:
removed falsely check in target dir

Ediacarium committed:
Some cleanup

Ediacarium committed:
Removed * dependency

Ediacarium committed:
Quick & Dirty autobind of all resettable PCI Devices

Ediacarium committed:
wait for vfio-ubind on shudown

Ediacarium committed:
Restricted execution to root or vfio-group members

Ediacarium committed:
The configuration now separates between permanent and temporary devices

This is for Bug #105
If you config before looked like:
  vfio_slots:
       "0000:02:ab.0"
       "0000:02:ab.1"
it now needs to looke like:
  vfio_slots:
    -
     Permanent:
       "0000:02:ab.0"
    -
     Permanent:
       "0000:02:ab.1"

Ediacarium committed:
Adapted wizard to allow adding aribtrary Vfio Devices for #105

Ediacarium committed:
Some Cleanup (removed the commented out previous version and cleaned up the output)

Ediacarium committed:
Rewrote the wizard to use udev instead of sysfs

Ediacarium committed:
Cleanup and fixed PCI Bridges shown in permanent PCI Selection

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	} else { 
+		info!("writing {} into {}", content, path.display());
+	
+		let mut file = OpenOptions::new().write(true).open(&path).expect(&format!("Failed to open {}", path.display()));
```
`File::create`?

main-- replied:
Not 100% equivalent:

>This function will create a file if it does not exist, and will truncate it if it does.

We don't really care about the truncation aspect but we definitely don't want to try to create random files in sysfs (not like it would work anyways but it doesn't make sense to even try).

oberien replied:
Sounds like a valid point ;)

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+use argparse::{ArgumentParser, StoreTrue, Store};
+use std::fs::{OpenOptions, read_link};
+use std::io::prelude::*;
+use users::{get_user_by_uid, get_effective_uid, get_current_uid, get_group_by_name};
```
Please don't import functions by-name, but use them with their module prefix: `users::get_user_by_uid`

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+		let user = get_user_by_uid(get_current_uid()).unwrap();
+		let user_name = user.name();
+		
+		if let None = vfio_group.members().iter().position(|gm| gm == user_name) {
```
`if !foo.any(condition)`

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+			warn!("No PCI domain supplied, assuming PCI domain is 0000");
+			device = "0000:".to_string() + &device;
+		}
+	}
```
`if !dbdf_regex.is_match(&device) && !bdf_regex.is_match(&device) { ... } else { ... }`

Ediacarium replied:
!dbdf_regex.is_match(&device) && !bdf_regex.is_match(&device) === true

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	if ! dev_reset.exists() {
+		error!("The device does no support resetting!");
+		info!("File {} didn't exist", dev_reset.display());
+		return;
```
This is an error. Shouldn't we panic instead of exiting without error code?

oberien commented on windows-gaming-driver/src/config.rs:
```diff
@@ -241,6 +241,21 @@ pub struct StorageDevice {
 #[derive(Deserialize, Serialize, Debug, Clone)]
+pub enum VfioDevice {
+	Permanent(String),
+	Temporarily(String)
```
If you call it `Permanent`, it should be called `Temporary`.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -25,9 +30,11 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
     println!("Success!");
     println!();
+    
+    println!("");
```
No need for the empty line above the print and the empty string inside `println! `

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -113,7 +113,24 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+				match child.wait() {
+					Ok(status) => 
+						if !status.success() {	
+						error!("vfio-ubind failed with {}! The device might not be bound to the vfio-driver and therefor not functional", status);
```
"and therefore not function correctly"

oberien commented on windows-gaming-driver/src/config.rs:
```diff
@@ -241,6 +241,21 @@ pub struct StorageDevice {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        f.write_str(match self {
+            &VfioDevice::Permanent(_) => "Permanent",
+            &VfioDevice::Temporarily(_) => "Temporarily",
```
~

main-- replied:
This is not a useful `Display` impl.

oberien commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,22 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+						error!("vfio-ubind failed with {}! The device might still be bound to the vfio-driver!", status);
+						},
+				Err(err) => error!("failed to wait on child. Got: {}", err)
+				
```
unnecessary empty line

oberien commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,22 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+				
+			}
+		}
+	}
```
The chance that we panic is pretty high. If we do panic during `core.run`, this cleanup code is not executed. Is it possible that you create a struct, which executes this code in the `Drop` implementation, so it'll be executed when unwinding?

main-- replied:
>The chance that we panic is pretty high.

Is it? If so, why? Note that this was restructured to no longer panic when an error is returned.

oberien replied:
At least I always get something like `thread 'main' panicked at 'Unexpected error: Connection reset by peer (os error 104)', src/driver/mod.rs:101` when I shutdown Windows.

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -113,7 +113,24 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+						error!("vfio-ubind failed with {}! The device might not be bound to the vfio-driver and therefor not functional", status);
+						},
+					Err(err) => error!("failed to wait on child. Got: {}", err)
+					
```
unneseccary empty line

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -113,7 +113,24 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+				device
+			}
+    		&VfioDevice::Permanent(ref device) => 
+    			device
```
Can be in the above line behind `=>`

main-- replied:
A match where both arms return the same thing generally indicates that you're doing it wrong. But this is not really relevant since that thing sholdn't be an enum to begin with.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	// filter to the display controller class (0x03XXXX, udev drops the leading zero)
+	let (gpu_devices, gpu_dev_ids, pci_devs) = ask_selection("Which graphics card do you want to pass through?", &|x| x.pci_class.starts_with("3") && x.pci_class.len() == 5, pci_devs, true);
+	
+	println!("Add aditional  pci devices(1/2)");
```
One space in the middle too much, one between "devices" and the opening parethesis too less.

oberien commented on windows-gaming-driver/src/util.rs:
```diff
@@ -23,3 +23,23 @@ pub fn usable_ports(bus: UsbBus) -> usize {
+			if filter(&dev) {
+				matched.push(dev);
+			}
+			else {
```
welp

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	
+	println!("Add aditional  pci devices(1/2)");
+	println!();
+	println!("Please choose aditional pci devices. These will be passed throught to qemu on boot.");
```
"through"

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	let (mut related_devices, unpassed_pci_devices) = devices.into_two(&|dev| dev.pci_device() == selected.pci_device());
+	related_devices.push(selected);
+	
+	let (mut ret_devices, mut ret_devids, ret_pci_devices) = ask_selection(question, filter, unpassed_pci_devices, permanent);
```
Recursion instead of a loop? That never ended well…

main-- replied:
Seconded. The recursion makes this **very** hard to read and I'm pretty sure restructuring this to a loop is actually trivial, right?

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!();
+	println!("Please choose aditional pci devices. These will be passed throught to qemu on boot.");
+	println!("NOTE: The devices listed here can be reset. They will be bound to the vfio-pci driver on qemu's start and unbound when it quits");
+	println!("This should usually work. If it doesn't ur on ur own.");
```
We should be formal and not super colloquial :)

Ediacarium replied:
As formal as this?
```
machine.memory = ask::anything("How much memory would you like to assign to it?",
                                       "be careful no validation LUL", |x| Some(x.to_owned()));
```
\- wm.rs ~line 67

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("If the device you want to pass through is not resettable it will show up in the next step.");
+	println!("If you want a device permanently bound to vfio-pci for some reason select it in the next step.");
+	println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+	println!("Helpful tools to avoid this and figure out which numbers are what devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
```
"which numbers are which devices"

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+	println!("Helpful tools to avoid this and figure out which numbers are what devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+	println!();
+	let (temp_devices, _, pci_devs) = ask_selection("Which PCI Devices do you want to pass through temporarily?", &|x| Path::new(PCI_DEVS_PATH).join(x.pci_slot.clone()).join("reset").exists(), pci_devs, false);
```
"would you like to"

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	let (temp_devices, _, pci_devs) = ask_selection("Which PCI Devices do you want to pass through temporarily?", &|x| Path::new(PCI_DEVS_PATH).join(x.pci_slot.clone()).join("reset").exists(), pci_devs, false);
+	println!("Add aditional pci devices(2/2)");
+	println!("");
+	println!("If you don't know what you are doing, chose None here!");
```
choose

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("U done goofed m8.");
+	println!("Ur on ur own :)");
+	println!();
+	//filter to everything except the PCI Bridges
```
Space between the comment-slashes and the content to make it more distiguishable from outcommented code.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("Ur on ur own :)");
+	println!();
+	//filter to everything except the PCI Bridges
+	let (perm_devices, perm_dev_ids, _) = ask_selection("Which PCI Device do you want to pass through permanently?", &|x| x.pci_class != "60400", pci_devs, true);
```
"would you like to"

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+
+    
+	// filter to the display controller class (0x03XXXX, udev drops the leading zero)
+	let (gpu_devices, gpu_dev_ids, pci_devs) = ask_selection("Which graphics card do you want to pass through?", &|x| x.pci_class.starts_with("3") && x.pci_class.len() == 5, pci_devs, true);
```
"would you like to"

oberien reviewed (changes requested):
¡4 spaces, not tabs!

main-- commented on Makefile:
```diff
@@ -14,6 +15,9 @@ test:
 	cd windows-gaming-driver && cargo build --release --locked
 
+$(VFIO_UBIND):
+	cd vfio-ubind && cargo build --release --locked
```
Consider using a cargo workspace for all our rust stuff.

(not 100% if that's the right solution for our case but I think it might be)

Ediacarium replied:
See #110 for this

main-- replied:
So is this postponed now or what? It's just that the major reason why I would prefer doing it with workspaces is that we bridge make to cargo by always invoking cargo which gets weird/slow fast when we have 10 cargo invocations every time.

oberien replied:
Postponed with #110

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+
+fn pretty_write(path: &Path, content : &str, dryrun: bool) {
+	if dryrun {
+		println!("writing {} into {}", content, path.display());
```
I have never seen dry-run affecting the actual semantics and I don't see why it ever should. If you actually are stupid enough to do dry-run with logging disabled, that's your problem.

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+		ap.refer(&mut dryrun).add_option(&["-d", "--dry-run"], StoreTrue, "Don't change anything");
+		ap.refer(&mut remove).add_option(&["-r", "--remove"], StoreTrue, "Reattach the default Driver to the device");
+		ap.parse_args_or_exit();
+	}
```
:+1:

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	
+	env_logger::init().unwrap();
+	
+	info!("effective uid: {} current uid: {}", get_effective_uid(), get_current_uid());
```
Consider downgrading this to debug

update: you seem to be relying on the fact that info is not displayed by default (thus using it as something like debug). in that case it's fine

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	info!("effective uid: {} current uid: {}", get_effective_uid(), get_current_uid());
+	
+	if get_effective_uid() != 0 {
+		error!("This tool has no root permission. Is the setuid bit not set or do you need to execute this as root?");
```
>Is the setuid bit not set or do you need to execute this as root?

This doesn't really make sense. How about: "This tool requires root permissions. If the setuid bit is not set, you need to execute this as root!"

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+
+	debug!("Checking with Regexes");
+	let dbdf_regex = Regex::new(r"^[[:xdigit:]]{4}:[[:xdigit:]]{2}:[[:xdigit:]]{2}.[[:xdigit:]]$").unwrap();
+	let bdf_regex = Regex::new(r"^[[:xdigit:]]{2}:[[:xdigit:]]{2}.[[:xdigit:]]$").unwrap();
```
I remain convinced that using regexes here is a bad idea. The entire reason why this is a separate binary is that we want to avoid giving full root access 20 untrusted libraries, so we should only use what we really need. From that perspective log+env_logger are already hard to justify but at least they are useful (debugging).

Regexes however give you nothing at all (as we discussed), you get the exact same behavior with something like this:

```rust
if device[2] == ':' {
    "0000:".to_owned() + device
} else {
    device
}
```

The important point here is that you're not actually using the regexes at all (to get their captured values). You are only checking if they match or not. But the system already implicitly does that for you because the sysfs folder won't exist if the ID is invalid.

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	let dev_sysfs = Path::new("/sys/bus/pci/devices/").join(&device);
+	let dev_iommu = dev_sysfs.join("iommu");
+	
+	if !dev_iommu.exists() {
```
This check is not necessary. While rebinding devices to vfio-pci is not very useful when iommu is off, I don't see any reason to prohibit it. Again, **this binary is not user-facing**!

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	
+	let dev_reset = dev_sysfs.join("reset");
+	
+	if ! dev_reset.exists() {
```
stray space

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+		let driver = if remove {"\n"} else {"vfio-pci"};
+		pretty_write(&dev_sysfs.join("driver_override"), driver, dryrun);
+		pretty_write(&dev_sysfs.join("driver/unbind"), &device, dryrun);		
+	}
```
```rust
} else if !remove {
    // must set this in this case as well
    pretty_write(&dev_sysfs.join("driver_override"), "vfio-pci", dryrun);
}
```

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+use users::{get_user_by_uid, get_effective_uid, get_current_uid, get_group_by_name};
+use users::os::unix::GroupExt;
+
+fn pretty_write(path: &Path, content : &str, dryrun: bool) {
```
Conventionally functions like this would be declared like this:

```rust
fn pretty_write<P: AsRef<Path>>(path: T, content: &str, dryrun: bool) {
```

This allows you to use `&Path` or `PathBuf` as parameter, depending on what you have.

(btw stray space in `content :`)

oberien replied:
It also allows `String`, `str`, `OsStr`, `OsString` and many other :)

main-- commented on windows-gaming-driver/src/config.rs:
```diff
@@ -218,7 +218,7 @@ pub struct MachineConfig {
 
     // convention: gpu is first
-    pub vfio_slots: Vec<String>,
+    pub vfio_slots: Vec<VfioDevice>,
```
:+1: 

(perhaps renaming this would be a good idea though now that this describes entire devices and not just slots)

main-- replied:
Actually, yeah: Please rename this now, this is a great opportunity as you have to rewrite your config anyways to switch to the struct. (We can't justify doing it later)

main-- commented on windows-gaming-driver/src/config.rs:
```diff
@@ -241,6 +241,21 @@ pub struct StorageDevice {
 }
 
 #[derive(Deserialize, Serialize, Debug, Clone)]
+pub enum VfioDevice {
```
I don't like this. How about:

```rust
struct VfioDevice {
    resettable: bool,
    slot: String,
    device: PciId,
}
```

(Device id is interesting anyways for the potential sanity checks on startup I mentioned. It also solves the issue where the setup section would store its own list of device IDs that can get out of sync/order with the slots in here)

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -17,12 +17,13 @@ use std::os::unix::net::{UnixListener as StdUnixListener};
 use std::fs::{self, Permissions};
 use std::os::unix::fs::PermissionsExt;
 use std::path::Path;
+use std::process::{Command};
```
Don't use braces if you're only importing one thing.

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,22 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
-
+    core.run(qemu.join(joined.or_else(|x| { error!("Unexpected error: {}", x); Ok(()) }))).expect("Unexpected error");
+
+	info!("unbinding resetable vfio-things");
```
typo: resettable

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,22 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+	
+    for slot in cfg.machine.vfio_slots.iter() {
+		if let &VfioDevice::Temporarily(ref device) = slot {
+			let mut child = Command::new(Path::new(::DATA_FOLDER).join("vfio-ubind")).arg(device).arg("-r").spawn().expect("failed to run vfio-ubind");
```
This function's `data` parameter is literally `Path::new(DATA_FOLDER)` (passed through from `main()`).

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,22 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+    for slot in cfg.machine.vfio_slots.iter() {
+		if let &VfioDevice::Temporarily(ref device) = slot {
+			let mut child = Command::new(Path::new(::DATA_FOLDER).join("vfio-ubind")).arg(device).arg("-r").spawn().expect("failed to run vfio-ubind");
+			match child.wait() {
```
I would just unwrap/expect here.

Ediacarium replied:
If we unwrap/expect here, we skip trying to unbind the remaining devices.

main-- commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -113,7 +113,24 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
-        qemu.args(&["-device", &format!("vfio-pci,host={},multifunction=on", slot)]);
+    	let vfio_device = match slot {
+			&VfioDevice::Temporarily(ref device) => {
+				let mut child = Command::new(Path::new(::DATA_FOLDER).join("vfio-ubind")).arg(device).spawn().expect("failed to run vfio-ubind");
```
~

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+	println!("Helpful tools to avoid this and figure out which numbers are what devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+	println!();
+	let (temp_devices, _, pci_devs) = ask_selection("Which PCI Devices do you want to pass through temporarily?", &|x| Path::new(PCI_DEVS_PATH).join(x.pci_slot.clone()).join("reset").exists(), pci_devs, false);
```
FYI udev devices have `syspath()`

this is very definitely the best way to access this path (check all the other udev code, that's why we don't have any existing "/sys/bus/pci" constants)

Now this *might* be hidden because we convert PCI devices to our own `PciDevice` struct but the fix in that case is to simply change `PciDevice` to carry the syspath.

main-- replied:
Actually come think of it, it's probably better to just check resettability in `PciDevice` directly.

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+	if !dbdf_regex.is_match(&device) {
+		if !bdf_regex.is_match(&device) {
+			println!("Please supply Domain:Bus:Device.Function of PCI device in form: dddd:bb:dd.f");
+			return;
```
This is an error. Shouldn't we panic to to make that clear?

main-- commented on windows-gaming-driver/src/util.rs:
```diff
@@ -23,3 +23,23 @@ pub fn usable_ports(bus: UsbBus) -> usize {
 }
 
+pub trait PrettySplit<T>{
+	fn into_two(self, &Fn(&T) -> (bool)) -> (Vec<T>, Vec<T>);
```
This already exists. It's called [partition](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition).

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    Ok(())	
+}
+
+fn ask_selection<'a>(question: &str, filter: &Fn(&PciDevice) -> (bool), pci_devices: Vec<PciDevice<'a>>, permanent: bool) -> (Vec<VfioDevice>, Vec<PciId>, Vec<PciDevice<'a>>) {
```
Rename this to something like "select_devices". The current name feels more like a generic ask function that belongs into the `ask` module and it doesn't indicate at all that this function actually loops until all selections are done.

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +86,93 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+	ret_devids.extend(related_devices.iter().map(|dev| dev.id));
+	
+	(ret_devices, ret_devids, ret_pci_devices)
+}
```
Please restructure this.

What you're doing right now:

* Ask for GPUs (or literally anything else)
* Ask for resettable devices (or literally anything else)
* Ask for nonresettable devices (or literally anything else)

(it should be obvious why this is bad)

What you should be doing:

* Ask for a *single* GPU (or *nothing at all*)
* Ask if they would like to add other PCI devices
* Only if they answered yes, ask for any number of PCI devices (but straight up prohibit bridges because wtf) (Here, we can initially hide non-resettable devices and only show them if they specifically ask for it, but warn in that case)

Ediacarium committed:
Commit for change request from pull-request
Fixed typos
Reworked the vfio-device selection process
Added more panics
TODO: Sanity check
TODO: refactor to workbench

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -65,14 +64,10 @@ fn main() {
-		}
+	match (dbdf_regex.is_match(&device), bdf_regex.is_match(&device)) {
+		(false, true) => {warn!("No PCI domain supplied, assuming PCI domain is 0000");	device = "0000:".to_string() + &device;},
+		(false, false) => {	panic!("Please supply Domain:Bus:Device.Function of PCI device in form: dddd:bb:dd.f");}
```
unnecessary tab between `{` and `panic`. In fact the whole block is not needed.

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -112,26 +112,21 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
-    	let vfio_device = match slot {
-			&VfioDevice::Temporarily(ref device) => {
-				let mut child = Command::new(Path::new(::DATA_FOLDER).join("vfio-ubind")).arg(device).spawn().expect("failed to run vfio-ubind");
+	//TODO: Check if the configured device is in the configured slot
```
Space between the comment-slashes and the content to make it more distiguishable from outcommented code.

Ediacarium committed:
Hopefully fixed the tab /whitespace thing

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -43,17 +42,17 @@ fn main() {
-		let vfio_group = get_group_by_name("vfio").expect("Your system has no vfio group. You need to be part of it to run this tool!");
-		let user = get_user_by_uid(get_current_uid()).unwrap();
+	if users::get_current_uid() != 0 {
+		let vfio_group = users::get_group_by_name("vfio").expect("Your system has no vfio group. You need to be part of it to run this tool!");
```
Newline before `.expect`, otherwise the line is pretty damn long.

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -65,14 +64,10 @@ fn main() {
-			device = "0000:".to_string() + &device;
-		}
+	match (dbdf_regex.is_match(&device), bdf_regex.is_match(&device)) {
+		(false, true) => {warn!("No PCI domain supplied, assuming PCI domain is 0000");	device = "0000:".to_string() + &device;},
```
Missing space between `{` and `warn`. Also if you open a block, put the content in new lines.

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -82,19 +77,18 @@ fn main() {
 		println!("Use intel_iommu=on or iommu=pt iommu=1");
 		info!("File {} didn't exist", dev_iommu.display());
-		return;
+		panic!();
```
Please include the error message into the `panic` instead of printing it normally beforehand.

oberien commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -99,18 +99,17 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
 				Ok(status) => 
 						if !status.success() {	
 						error!("vfio-ubind failed with {}! The device might still be bound to the vfio-driver!", status);
 						},
```
Use match guards here to include the if condition into the match arm. This does add a new match arm, but makes the code more readable.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -91,84 +88,78 @@ fn select(setup: &mut SetupConfig, machine: &mut MachineConfig) -> Result<()> {
+		println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+		println!("Helpful tools to avoid this and figure out which numbers are which devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+		println!();
+		let resetable_devices = pci_devs.iter().filter(|x| Path::new(PCI_DEVS_PATH).join(x.pci_slot.clone()).join("reset").exists()).collect();
```
You can check the `PciDevice.resettable` flag here.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -91,84 +88,78 @@ fn select(setup: &mut SetupConfig, machine: &mut MachineConfig) -> Result<()> {
 	
-	(ret_devices, ret_devids, ret_pci_devices)
+	if selection < askable_devices.len() {
+		let ref selected = askable_devices[selection];
```
Does `let selected = &askable_devices[selection]` work here?

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -91,84 +88,78 @@ fn select(setup: &mut SetupConfig, machine: &mut MachineConfig) -> Result<()> {
+			};
+			
+			if !device.resettable {
+				setup.vfio_devs.push(device.id);
```
Can't we iterate through `machine.vfio_slots` where we are currently using `setup.vfio_devs`?

Ediacarium replied:
In this code we can, however at least iommu.rs needs it later on.

You *could* argue that the vfio-module should just do the iommu checking directly instead of putting it into a different step, which could help, if you already have iommu enabled for some reason and we could've checked it when you first select the vfio-devices. But i think that is an entirely different issue.

oberien reviewed (changes requested):
¡4 spaces instead of tabs!

Ediacarium committed:
Fixed annotations from pull request
Fixed panic without panic message
Improved redability

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -65,19 +66,19 @@ fn main() {
-        (false, true) => {warn!("No PCI domain supplied, assuming PCI domain is 0000");	device = "0000:".to_string() + &device;},
-        (false, false) => {	panic!("Please supply Domain:Bus:Device.Function of PCI device in form: dddd:bb:dd.f");}
+        (false, true) => { 
+            warn!("No PCI domain supplied, assuming PCI domain is 0000");	device = "0000:".to_string() + &device; },
```
really¿¿¿

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -65,19 +66,19 @@ fn main() {
+        (false, true) => { 
+            warn!("No PCI domain supplied, assuming PCI domain is 0000");	device = "0000:".to_string() + &device; },
+        (false, false) => {	
+            panic!("Please supply Domain:Bus:Device.Function of PCI device in form: dddd:bb:dd.f"); }
```
This still doesn't need a block.

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -65,19 +66,19 @@ fn main() {
-        println!("Use intel_iommu=on or iommu=pt iommu=1");
         info!("File {} didn't exist", dev_iommu.display());
-        panic!();
+        panic!("No signs of an IOMMU. Check your hardware and/or linux cmdline parameters.Use intel_iommu=on or iommu=pt iommu=1");
```
The line is too long. Maybe split the string into multiple lines.

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -112,16 +112,15 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
-                        },
-                    Err(err) => panic!("failed to wait on child. Got: {}", err)	
+                    Ok(status) if !status.success() => 
+                        panic!("vfio-ubind failed with {}! The device might not be bound to the vfio-driver and therefore not function correctly", status),
```
~

oberien reviewed (commented):
Finally spaces ![](https://static-cdn.jtvnw.net/emoticons/v1/41/1.0)

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -91,84 +88,78 @@ fn select(setup: &mut SetupConfig, machine: &mut MachineConfig) -> Result<()> {
+			};
+			
+			if !device.resettable {
+				setup.vfio_devs.push(device.id);
```
Reference #108

Ediacarium committed:
General prettyfication

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -118,7 +118,8 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
                     Ok(status) if !status.success() => 
-                        panic!("vfio-ubind failed with {}! The device might not be bound to the vfio-driver and therefore not function correctly", status),
+                        panic!("vfio-ubind failed with {}! \
+                                The device might not be bound to the vfio-driver and therefore not function correctly", status),
```
still too long. Sorry m8.

Ediacarium committed:
Smol lines

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -119,7 +119,11 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+                                bound to the \ 
+                                vfio-driver and \
+                                therefore not function \
+                                correctly", status),
```
I'll leave this up to @main--

oberien commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,21 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+    
+    for dev in cfg.machine.vfio_slots.iter() {
+        if dev.resettable {
+            let mut child = Command::new(data.join("vfio-ubind")).arg(&dev.slot).arg("-r").spawn().expect("failed to run vfio-ubind");
```
I think you should plit this up into one line per function invocation.

oberien commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,21 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+            let mut child = Command::new(data.join("vfio-ubind")).arg(&dev.slot).arg("-r").spawn().expect("failed to run vfio-ubind");
+            match child.wait() {
+                Ok(status) => 
+                        if !status.success() {	
```
Here again: inline the condition as match guard.

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -112,9 +112,25 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+    // TODO: Check if the configured device is in the configured slot
+    for device in cfg.machine.vfio_slots.iter() {
+        if device.resettable {
+                let mut child = Command::new(data.join("vfio-ubind")).arg(&device.slot).spawn().expect("failed to run vfio-ubind");
```
I think you should plit this up into one line per function invocation.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -2,11 +2,14 @@ use std::borrow::Cow;
 use setup::wizard;
+use pci_device::PciDevice;
+
+
```
3 empty lines are 2 2 much

main-- replied:
???

oberien replied:
Still a problem.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    // filter to the display controller class (0x03XXXX, udev drops the leading zero)
+    let gpus = pci_devs.iter().filter(|x| x.pci_class.starts_with("3") && x.pci_class.len() == 5).collect();
+    select_device(setup, machine, "Which graphics card would you like to pass through?", &gpus)?;
+    if ask::yesno("Do you want to pass aditional pci devices?") {
```
"Would you like to"

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!();
+        println!("Please choose aditional pci devices. These will be passed through to qemu on boot.");
+        println!("NOTE: The devices listed here can be reset. They will be bound to the vfio-pci driver on qemu's start and unbound when it quits");
+        println!("This should usually work. If it doesn't you're on your own.");
```
"If it doesn't, you'll be on your own" - get conditional clauses right ;)

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("NOTE: The devices listed here can be reset. They will be bound to the vfio-pci driver on qemu's start and unbound when it quits");
+        println!("This should usually work. If it doesn't you're on your own.");
+        println!("You will not be able to use these devices while qemu is running.");
+        println!("If the device you want to pass through is not resettable (like a gpu) it will show up in the next step.");
```
`like a gpu` - But you already selected the GPU, so it won't show up in the next menu. Maybe find a better example here.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("This should usually work. If it doesn't you're on your own.");
+        println!("You will not be able to use these devices while qemu is running.");
+        println!("If the device you want to pass through is not resettable (like a gpu) it will show up in the next step.");
+        println!("If you want a device permanently bound to vfio-pci for some reason select it in the next step.");
```
"for some reason, you'll need to" - get your conditional clauses right ;) Also, it'll be too long for a single line then.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("If the device you want to pass through is not resettable (like a gpu) it will show up in the next step.");
+        println!("If you want a device permanently bound to vfio-pci for some reason select it in the next step.");
+        println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+        println!("Helpful tools to avoid this and figure out which numbers are which devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
```
Split into multiple lines.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("Helpful tools to avoid this and figure out which numbers are which devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+        println!();
+        let resetable_devices = pci_devs.iter().filter(|x| x.resettable).collect();
+        while let Ok(Some(())) = select_device(setup, machine, "Which device do you want to pass through?", &resetable_devices) {};
```
"would you like to"

main-- replied:
`Option<()>` is `bool`. Just use the latter. Also, you're not supposed these errors - they come straight from udev and are thus fatal. Just do this:

```rust
while select_device(setup, machine, "...", &resettable_devices).expect("...") { }
```

(Also, the trailing `;` is wrong.)

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("If you don't know what you are doing, choose None here!");
+        println!("NOTE: EVERY PCI DEVICE LISTED HERE WILL BE PERMANENTLY BOUND TO VFIO-PCI!");
+        println!("ONLY ADD DEVICES HERE IF YOU KNOW WHAT YOU ARE DOING!");
+        println!("If you kill your system using this there is only one thing i can say to you:");
```
"using this, there'll only be" - get your conditional clauses right ;)

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("Ur on ur own :)");
+        println!();
+        let all_devices = pci_devs.iter().filter(|x| x.pci_class != "60400").collect();
+        while let Ok(Some(())) = select_device(setup, machine, "Which device do you want to pass through?", &all_devices) {};
```
"would you like to"

main-- replied:
~

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    iter.match_subsystem("pci")?;
+    let udev_devices = iter.scan_devices()?.map(PciDevice::new);
+    let pci_devs: Vec<_> = udev_devices.filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
+    let askable_devices: Vec<_> = devices.iter().filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
```
Split into multiple lines. Each function (except for `.iter()`) into its own line.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        }
+        Ok(Some(()))
+    }
+    else{
```
welp

oberien commented on windows-gaming-driver/src/util.rs:
```diff
@@ -22,4 +22,3 @@ pub fn usable_ports(bus: UsbBus) -> usize {
         UsbBus::Xhci => 15,
     }
 }
-
```
?

oberien replied:
Still an issue

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+    if dryrun {
+        println!("writing {} into {}", content, path.display());
+    } else { 
+        info!("writing {} into {}", content, path.display());
```
see old review (why print vs info?)

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+    debug!("effective uid: {} current uid: {}", users::get_effective_uid(), users::get_current_uid());
+    
+    if users::get_effective_uid() != 0 {
+        error!("This tool requires root permissions. If the setuid bit is not set, you need to execute this as root!");
```
No return/panic/anything? I'm sure you need to exit here.

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+        if !vfio_group.members().contains(&user_name.to_owned()) {
+            panic!("You're not part of the vfio group. Apply there and run this tool again!");
+        }
+        else {
```
style nit

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+        },
+        (false, false) => panic!("Please supply Domain:Bus:Device.Function of PCI device in form: dddd:bb:dd.f"),
+        (true, _) => (),
+    }
```
again, get rid of those damn regexes, they accomplish nothing

main-- commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+    
+    if let Ok(driver) = dev_driver {
+        if driver.file_name().unwrap() == "vfio-pci" && !remove {
+            println!("Device already bound to vfio-pci driver, nothing to do here");
```
Why suddenly print (instead of logging)?

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,21 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+    info!("unbinding resettable vfio-things");
+    
+    for dev in cfg.machine.vfio_slots.iter() {
+        if dev.resettable {
```
Please use `iter().filter()` instead of this `if`.

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        println!("MAKE SURE YOU DON'T PASS THROUGH YOUR USB-CONTROLLER TO WHICH YOUR KEYBOARD AND MOUSE IS CONNECTED!");
+        println!("Helpful tools to avoid this and figure out which numbers are which devices are lspci and lsusb with the -v, -t (lsusb only) and -nn (lspci only) command.");
+        println!();
+        let resetable_devices = pci_devs.iter().filter(|x| x.resettable).collect();
```
typo: resettable

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    let udev = Context::new().expect("Failed to create udev context");
+    let mut iter = Enumerator::new(&udev)?;
+    iter.match_subsystem("pci")?;
+    let udev_devices = iter.scan_devices()?.map(PciDevice::new);
```
???????????????????

ffs why are you re-querying these?

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+        let resetable_devices = pci_devs.iter().filter(|x| x.resettable).collect();
+        while let Ok(Some(())) = select_device(setup, machine, "Which device do you want to pass through?", &resetable_devices) {};
+        
+        println!("Add aditional pci devices (2/2)");
```
*sigh*

I told you to hide this and only show it to users that *specifically* request to see it.

Please.

Even if it's just another yesno around this thing.

Hide it.

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    Ok(())	
+}
+
+fn select_device(setup: &mut SetupConfig, machine: &mut MachineConfig, question: &str, devices: &Vec<&PciDevice>) -> Result<Option<()>>{
```
`&Vec<T>` is rarely useful, you should basically always use `&[T]` instead.

---

I wish you got rid of `setup.vfio_devs` in this PR instead of postponing it for someone else to do.

main-- commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,87 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+    let pci_devs: Vec<_> = udev_devices.filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
+    let askable_devices: Vec<_> = devices.iter().filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
+    
+    println!("{}", devices.len());
```
?

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -96,10 +97,21 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
-    match core.run(qemu.select(joined)) {
-        Ok(((), _)) => (), // one (*hopefully* qemu) is done, so the other is too
-        Err((e, _)) => panic!("Unexpected error: {}", e),
+    core.run(qemu.join(joined.or_else(|x| { error!("Unexpected error: {}", x); Ok(()) }))).expect("Unexpected error");
```
ijkolnufuahcdcjhnixckgfdxcgijkofgdcsuijohfcgduijohfurgdsijohfgcrdsijohl

main-- replied:
```rust
core.run(qemu.select2(joined).then(|x| {
    match x {
        Ok(Either::A((x, _))) => info!("qemu down first, all ok"),
        Err(Either::A((e, _))) => error!("..."),
        Ok(Either::B((_, _))) => unreachable!(), // we never return cleanly
        Err(Either::B((e, a))) => {
            error!("...");
            return a.boxed(); // we errored first, wait for qemu to exit
        }
    }
    future::ok(()).boxed()
)).expect("...");
```

main-- committed:
Merge branch 'master' into Dawn_of_the_VFIO_2.0_Episode_2_Reborn_Global_Offensive_Infinte_U_Vite_A_New_Hope_Remastered_Enhanced_Edition_Rebooted_electric_boogaloo_on_new_tides

main-- committed:
Address review to vfio-ubind

main-- committed:
Address review to windows-gaming-driver

main-- committed:
vfio-ubind: Sanitize input

Ediacarium commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -82,82 +82,69 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
-        for device in pci_devs.iter().filter(|x| x.pci_device() == selected.pci_device()) {
+
+    if let Some(selected_id) = askable_devices.get(selection).map(|x| x.id) {
+        for device in devices.iter().filter(|x| x.id == selected_id) {
```
This does not do what you think it does. This check for the specific id. However the graphics and audio part of a gpu have different ids: 
> NVIDIA Corporation Unknown model [10de:1b06]
> NVIDIA Corporation GP102 HDMI Audio Controller [10de:10ef]

Ediacarium committed:
Dirty fixed vfio-wizard & minor splling mistake

Ediacarium commented on vfio-ubind/src/main.rs:
```diff
@@ -57,8 +57,10 @@ fn main() {
+    let devices_folder = Path::new("/sys/bus/pci/devices/");
+    let dev_sysfs = devices_folder.join(&device).canonicalize()
+        .expect("Failed to look up the given device (does it exist?)!");
+    assert_eq!(dev_sysfs.parent(), Some(devices_folder));
```
> windows-gaming-driver[4091]: thread 'main' panicked at 'assertion failed: `(left == right)` (left: `Some("/sys/devices/pci0000:00/0000:00:1c.4")`, right: `Some("/sys/bus/pci/devices/")`)', src/main.rs:63

main-- replied:
we lost

main-- committed:
Finally fix the vfio-ubind input sanitization

oberien commented on vfio-ubind/src/main.rs:
```diff
@@ -0,0 +1,97 @@
+        info!("File {} didn't exist", dev_iommu.display());
+        panic!("No signs of an IOMMU. \
+                Check your hardware and/or linux cmdline parameters. \
+                Use intel_iommu=on or iommu=pt iommu=1");
```
```
Use `intel_iommu=on` or `iommu=pt iommu=1`"
```

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -115,9 +115,25 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+    // TODO: Check if the configured device is in the configured slot
+    for device in cfg.machine.pci_devices.iter() {
+        if device.resettable {
+                let mut child = Command::new(data.join("vfio-ubind")).arg(&device.slot).spawn().expect("failed to run vfio-ubind");
```
Indentation ![](https://static-cdn.jtvnw.net/emoticons/v1/28087/1.0)

oberien commented on windows-gaming-driver/src/driver/qemu.rs:
```diff
@@ -115,9 +115,25 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path, clientpipe_path: &Path, monito
+                                bound to the \
+                                vfio-driver and \
+                                therefore not function \
+                                correctly", status),
```
¯\_(ツ)_/¯

oberien commented on windows-gaming-driver/src/config.rs:
```diff
@@ -252,6 +250,13 @@ pub struct StorageDevice {
+pub struct VfioDevice {
+    pub resettable: bool,
+    pub slot: String,
+    pub device: PciId,
```
`VfioDevice.device` feels weird, I guess `VfioDevice.id` would be better.

Ediacarium replied:
This is what it would currently look like in the config:
(Assuming github would respect the indentation)

>   pci_devices: 
>     - 
>       resettable: false
>       slot: "0000:02:00.0"
>       device: 
>         vendor: 4318
>         device: 6918

Having 

>   pci_devices: 
>     - 
>       resettable: false
>       slot: "0000:02:00.0"
>       id: 
>         vendor: 4318
>         device: 6918

feels strange to me as well

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,75 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+fn select_device<P: Fn(&PciDevice) -> bool>(machine: &mut MachineConfig,
+                                            question: &str,
+                                            devices: &[PciDevice],
+                                            predicate: P) -> Result<bool> {
```
Use a `where` here and put multiple arguments in the same line.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,75 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+
+    if let Some(selected_slot) = askable_devices.get(selection).map(|x| x.pci_slot.clone()) {
+        for device in devices.iter()
+            .filter(|x| x.pci_slot[..x.pci_slot.len()-1] == selected_slot[..selected_slot.len()-1]) {
```
`selected_slot[..selected_slot.len()-1]` is the same as `selected_slot[..]`. Then it can also be all in the same line.

oberien replied:
Actually it's not. It strips away the last byte.

oberien commented on windows-gaming-driver/src/setup/vfio.rs:
```diff
@@ -79,6 +82,75 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
+                  |x| x.pci_class.starts_with("3") && x.pci_class.len() == 5)?;
+    if ask::yesno("Would you like to pass aditional PCI devices?") {
+        println!();
+        println!("Please choose aditional pci devices. These will be passed through to the VM on boot.");
```
`on boot` is ambiguous as it could mean both windows boot and linux boot.

Ediacarium replied:
Well how should something be passed to the vm on (linux) boot when the vm is not running?

Although we could indeed word it better.

oberien committed:
Fix some style

Ediacarium committed:
Removed stray debug message

main-- commented on vfio-ubind/.gitignore:
```diff
@@ -0,0 +1,10 @@
+# These are backup files generated by rustfmt
+**/*.rs.bk
+
+# vim swap files
```
might wanna get this shit out of the way one day (not lang-dependent so I moved it to the root gitignore)

main-- commented on windows-gaming-driver/src/driver/mod.rs:
```diff
@@ -138,10 +139,28 @@ pub fn run(cfg: &Config, tmp: &Path, data: &Path) {
+    core.run(qemu.select2(joined).then(|x| {
+        match x {
+            Ok(future::Either::A((_, _))) => info!("qemu down first, all ok"),
+            Err(future::Either::A((e, _))) => return future::err(e).boxed(),//error!("Waiting for qemu errored: {}", e),
```
wtf did I commit this???

main-- reviewed (changes requested):
the one comment doesn't belong there, rest looking great

main-- committed:
Fix remaining nits
